### PR TITLE
"Fix" acceptance tests

### PR DESCRIPTION
**Pull request by plexus**
It seems there were a couple of problems with the acceptance test, the biggest being that the `let()` blocks defined in shared example groups were interfering with each other. To prevent this I've wrapped them all in `context {}` blocks. Now the tests break though, because there were some issues going unnotticed.
    
Also added a new "list of quotes" fixture to test/document the behavior of rendering a top level collection, and renamed the fixtures for collection json to use the right naming convention.

We already exposed several of the pipeline steps as isolated methods, this adds "format" to that list, so you can turn a Yaks::Resource into a formatted data structure targeting a specific output format, without serializing it to string.

The files in yaks/spec/json/ are serialized representations of the data loaded from yaks/spec/yaml/. The JSON-API version had drifted from the defined data, and because our acceptance tests were broken we did not catch it.

Occasionally I drop a binding.pry in a test to inspect some values, in that case I don't want to have it time out immediately. In stead of commenting out the relevant section each time (and accidentally committing that :P) I can now do `NO_TIMEOUT=1 bundle exec rspec`

### CODE REVIEW

**Comment by plexus** [comment]
Anyone like to do a code review on this? I'd like to merge so we can in a next step improve CJ's handling of self links (issue #87), and bring the JSON-API reader in line with its writer.

And really I just want to write the Siren formatter. Guess why it's called Yaks :stuck_out_tongue: 

/ping @groyoh @janko-m @chastell

**Comment by groyoh** [comment]
I can review this later today :wink:

**Comment by janko-m** [comment]
This looks really good. It's great that you started each shared examples with `describe` block, so that you don't have to have those empty `context { ... }`. It's also great that you added `NO_TIMEOUT`, I also missed being able to `binding.pry` easily. How come you set timeouts in the first place?

Just one very slight opinion, I think we can ingore some Rubocop complaints, for example `eql(foo: "bar")` instead of `eql({foo: "bar"})`. In the context of JSON testing I prefer the second version, because we semantically are testing hashes and not options. The same with some other Rubocop complaints you may not agree on.

Is Siren good?

File: yaks/spec/json/confucius.json_api.json [hunk 1]
```diff
@@ -1,38 +1,18 @@
 {
-  "meta": {
-    "page": {
-      "limit": 10,
-      "offset": 0,
-      "total": 2
+  "data": {
+    "id": 9,
```
Comment (groyoh) [inline yaks/spec/json/confucius.json_api.json]: This is not really related to your PR itself but JSON API ids must be string (http://jsonapi.org/format/#document-structure-resource-ids). I think we should create a task for this.

File: yaks/lib/yaks/reader/hal.rb [hunk 2]
```diff
@@ -29,7 +29,7 @@ def convert_links(links)
             rel = rel.to_sym if rel =~ /\A\w+\z/
             Resource::Link.new(rel: rel, uri: l['href'], options: options)
           end
-        end.to_set
+        end
```
Comment (groyoh) [inline yaks/lib/yaks/reader/hal.rb]: Why was it a set in the first place? I do not know much about hal.

File: yaks/spec/acceptance/json_shared_examples.rb [hunk 3]
```diff
@@ -1,19 +1,52 @@
...
+RSpec.shared_examples_for 'JSON Writer' do |fixture_name|
+  describe 'Yaks::Resource => JSON' do
+    let(:object)  { load_yaml_fixture(fixture_name) }
+    let(:json_fixture) { load_json_fixture("#{fixture_name}.#{format_name}") }
+    let(:serialized) {
+      yaks_config.call(object, hooks: [[:skip, :serialize]], format: format_name)
+    }
 
-  subject { yaks.call(input) }
+    # before do
```
Comment (groyoh) [inline yaks/spec/acceptance/json_shared_examples.rb]: Is this wanted or you forgot to remove it?

**Comment by groyoh** [comment]
One or two remarks, but it looks good to me! :+1:

@janko-m we could add this to the rubocop config:
```yml
Style/BracesAroundHashParameters:
  Exclude:
    - yaks/spec/*
```
so that it is only disabled for specs.

File: yaks/spec/json/confucius.json_api.json (hunk 1 above)
Comment (janko-m) [inline yaks/spec/json/confucius.json_api.json]: Good catch! I already added this to https://github.com/plexus/yaks/pull/86, which will get merged once I get changes from this PR :wink:

File: yaks/lib/yaks/reader/hal.rb (hunk 2 above)
Comment (plexus) [inline yaks/lib/yaks/reader/hal.rb]: This has nothing to do with HAL. The `links` on a `Yaks::Resource` could be considered conceptually a set, because the ordering is undefined (i.e. they might have a certain order but it's not something you should rely on). I guess when I wrote this code I had that on my mind. 

In early versions of Yaks I actually used [Hamster](https://github.com/hamstergem/hamster) data structures for these things, since most things in Yaks are immutable, but in the end decided to keep it simple and just use Ruby arrays and hashes.

File: yaks/spec/acceptance/json_shared_examples.rb (hunk 3 above)
Comment (plexus) [inline yaks/spec/acceptance/json_shared_examples.rb]: I left it in deliberately, found it can be quite helpful when debugging these tests.

**Comment by plexus** [comment]
> How come you set timeouts in the first place?

This is for Mutant. It is possible that a mutation results in an infinite loop, causing Mutant to hang. Actually we could probably enable the timeouts *only* when running mutant.

> Rubocop doesn't always know best

Totally. This is the first time I'm enforcing it on a full project, so I'm still in the phase of getting a feel for it. In general I don't always like tools telling me what to do, but I want to stick with it for a bit to find out what parts about it I like and don't like. In general I do really appreciate having consistent styling across a project.

In these tests I agree an explicit hash would be more clear. I'll add the exclusion as suggested.

> Is Siren good?

Yes :) From what I've seen it's the only format with links and forms that is really gaining traction. There are already a bunch of companies using it in production, and there's an active community around it. It's also a very good match for Yaks because its data model is quite close to `Yaks::Resource`. So I really want to get this in so I can reach out to that community.

File: yaks/lib/yaks/reader/hal.rb (hunk 2 above)
Comment (groyoh) [inline yaks/lib/yaks/reader/hal.rb]: Ok thanks for the explanation. So it has nothing to do with uniqueness of items in set?

Comment (plexus) [inline yaks/lib/yaks/reader/hal.rb]: To some extent, I don't see how it would make sense to have two identical links. But the ordering is something I've actually ran into.

Comment (groyoh) [inline yaks/lib/yaks/reader/hal.rb]: Ok, when I saw the `to_set` call I first thought about uniqueness rather than ordering since two identical links does not make sense as you said.

**Comment by groyoh** [comment]
@plexus I'm for enabling timeouts only with mutant :+1:

**Comment by chastell** [comment]
> This is the first time I'm enforcing [RuboCop] on a full project, so I'm still in the phase of getting a feel for it. In general I don't always like tools telling me what to do, but I want to stick with it for a bit to find out what parts about it I like and don't like. In general I do really appreciate having consistent styling across a project.

Oh definitely – I really like consistency, much more than agreeing with RuboCop’s defaults. :)

> In these tests I agree an explicit hash would be more clear. I'll add the exclusion as suggested.

You can [tell RuboCop to enforce braces](https://github.com/bbatsov/rubocop/blob/395eaf587b8d3e835760943d7d39714e7b84dafb/config/default.yml#L218-L230) in these cases.

**Comment by plexus** [comment]
Yeah but 9 times out of 10 I don't want the braces. It's only when explicitly testing for data that I do want them.

**Comment by plexus** [comment]
I guess that's it. Will wait for the build to finish and then merge.

File: .rubocop.yml [hunk 4]
```diff
@@ -138,3 +138,9 @@ Style/UnneededPercentQ:
 
 Style/TrailingComma:
   Enabled: false
+
+# Allow redundant braces in foo.bar({"qux" => "quz"}), when writing
+# e.g. JSON tests this can be more explicit and clear
+Style/BracesAroundHashParameters:
+  Exclude:
+    - yaks/spec/*
```
Comment (groyoh) [inline .rubocop.yml]: I made a mistake for this, the pattern should be `yaks/spec/**/*'.

Comment (plexus) [inline .rubocop.yml]: Hah, that'll teach me not to copy things blindly :)

\*blindly copies `yaks/spec/**/*`*

**Comment by plexus** [comment]
I have to say as much as I enjoy pushing to master, it's darn nice to be able to get some feedback :smile:
